tracer/src/Datadog.Trace.Tools.Analyzers/ConfigurationAnalyzers/EnvironmentGetEnvironmentVariableAnalyzer.cs

```diff
@@ -0,0 +1,255 @@
+            messageFormat: "Environment.GetEnvironmentVariable must use a constant from ConfigurationKeys or PlatformKeys class, not hardcoded string '{0}'",
+            category: "Usage",
+            defaultSeverity: DiagnosticSeverity.Error,
+            isEnabledByDefault: true,
+            description: "Use ConfigurationKeys.* or PlatformKeys.* constants instead of hardcoded strings to prevent typos.");
+
+        /// <summary>
+        /// Diagnostic descriptor for when GetEnvironmentVariable is called with anything other than a ConfigurationKeys or PlatformKeys constant.
+        /// </summary>
+        public static readonly DiagnosticDescriptor UseConfigurationKeysNotVariablesRule = new(
+            id: "DD0012",
+            title: "Use ConfigurationKeys or PlatformKeys constants in Environment.GetEnvironmentVariable calls",
+            messageFormat: "Environment.GetEnvironmentVariable must use a constant from ConfigurationKeys or PlatformKeys class, not '{0}'",
+            category: "Usage",
+            defaultSeverity: DiagnosticSeverity.Error,
+            isEnabledByDefault: true,
+            description: "Use ConfigurationKeys.* or PlatformKeys.* constants directly. Variables, parameters, and expressions are not allowed.");
+
+        /// <summary>
+        /// List of methods to check. Add new entries here to extend the analyzer.
+        /// Format: ("Fully.Qualified.TypeName", "MethodName")
+        /// </summary>
+        private static readonly (string TypeName, string MethodName)[] TargetMethods =
+        {
+            ("Datadog.Trace.Util.EnvironmentHelpers", "GetEnvironmentVariable"),
+            ("Datadog.Trace.Util.EnvironmentHelpers", "EnvironmentVariableExists"),
+            ("Datadog.Trace.Util.EnvironmentHelpersNoLogging", "GetEnvironmentVariable"),
+            ("Datadog.Trace.Util.EnvironmentHelpersNoLogging", "TryCheckEnvVar"),
+            ("Datadog.Trace.Ci.CiEnvironment.IValueProvider", "GetValue"),
+        };
+
+        /// <summary>
+        /// Gets the supported diagnostics
+        /// </summary>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(UseConfigurationKeysRule, UseConfigurationKeysNotVariablesRule);
+
+        /// <summary>
+        /// Initialize the analyzer
+        /// </summary>
+        /// <param name="context">context</param>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeInvocationExpression, SyntaxKind.InvocationExpression);
+        }
+
+        private static void AnalyzeInvocationExpression(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = (InvocationExpressionSyntax)context.Node;
+
+            // Get the method being called
```

Comment:
nit: getting the semantic model is _comparatively_ expensive, and this is going to get called for every single node in the compilation. We should do everything we can to bail out asap, as cheaply as possible.

In particular, I would hardcode some initial name checks which will act as a good filter.  You _could_ also check based on argument lengths etc. Basically, anything you can reliably add to avoid using the SemanticModel is likely going to be a net win:


```suggestion
        if (invocation.Expression is not MemberAccessExpressionSyntax memberAccess
            || (memberAccess.Name.Identifier.Text is not "GetEnvironmentVariable" or "EnvironmentVariableExists" or "TryCheckEnvVar" or "GetValue")
        {
            return;
        }

            // Get the method being called
```